runtime/runtime/src/actions.rs

```diff
@@ -624,6 +627,202 @@ pub(crate) fn action_add_key(
+    result.gas_used = safe_add_gas(result.gas_used, prepaid_send_fees)?;
+    result.gas_burnt = safe_add_gas(result.gas_burnt, prepaid_send_fees)?;
+    result.new_receipts.push(new_receipt);
+
+    Ok(())
+}
+
+fn receipt_required_gas(apply_state: &ApplyState, receipt: &Receipt) -> Result<Gas, RuntimeError> {
+    Ok(match &receipt.receipt {
+        ReceiptEnum::Action(action_receipt) => {
+            let mut required_gas = safe_add_gas(
+                total_prepaid_exec_fees(
+                    &apply_state.config.transaction_costs,
+                    &action_receipt.actions,
+                    &receipt.receiver_id,
+                    apply_state.current_protocol_version,
+                )?,
+                total_prepaid_gas(&action_receipt.actions)?,
+            )?;
+            required_gas = safe_add_gas(
+                required_gas,
+                apply_state.config.transaction_costs.action_receipt_creation_config.exec_fee(),
+            )?;
+
+            required_gas
+        }
+        ReceiptEnum::Data(_) => 0,
+    })
+}
+
+fn validate_delegate_action_key(
+    state_update: &mut TrieUpdate,
+    apply_state: &ApplyState,
+    delegate_action: &DelegateAction,
+    result: &mut ActionResult,
+) -> Result<(), RuntimeError> {
+    // 'delegate_action.sender_id' account existence must be checked by a caller
+    let mut access_key = match get_access_key(
+        state_update,
+        &delegate_action.sender_id,
+        &delegate_action.public_key,
+    )? {
+        Some(access_key) => access_key,
+        None => {
+            result.result = Err(ActionErrorKind::DelegateActionAccessKeyError(
+                InvalidAccessKeyError::AccessKeyNotFound {
+                    account_id: delegate_action.sender_id.clone(),
+                    public_key: delegate_action.public_key.clone(),
+                },
+            )
+            .into());
+            return Ok(());
+        }
+    };
+
+    if delegate_action.nonce <= access_key.nonce {
+        result.result = Err(ActionErrorKind::DelegateActionInvalidNonce {
+            delegate_nonce: delegate_action.nonce,
+            ak_nonce: access_key.nonce,
+        }
+        .into());
+        return Ok(());
+    }
+
+    let upper_bound = apply_state.block_height
+        * near_primitives::account::AccessKey::ACCESS_KEY_NONCE_RANGE_MULTIPLIER;
+    if delegate_action.nonce >= upper_bound {
+        result.result = Err(ActionErrorKind::DelegateActionNonceTooLarge {
+            delegate_nonce: delegate_action.nonce,
+            upper_bound,
+        }
+        .into());
+        return Ok(());
+    }
+
+    access_key.nonce = delegate_action.nonce;
+
+    let actions = delegate_action.get_actions();
+
+    if let AccessKeyPermission::FunctionCall(ref function_call_permission) = access_key.permission {
+        if actions.len() != 1 {
+            result.result = Err(ActionErrorKind::DelegateActionAccessKeyError(
```

Comment:
I'm confused -- if there is more than 1 action we return 'invalid access key'? Can you explain in a comment?

Reply:
This part of code is similar to the one in `verify_and_charge_transaction`. I hope this behavior is documented somewhere (access key permissions) because I can't find this in the documentation :)  This is a restriction of "function call" access keys. The transaction must contain exactly one `FunctionCall` if "function call" access key is used